node/lib/packaging/LocalPackageSource.ts

```diff
@@ -0,0 +1,136 @@
+
+import { PackageInfo } from "./PackageInfo";
+import { PackageSource } from "./PackageSource";
+
+import * as fs from "mz/fs";
+import * as path from "path";
+
+/**
+ * Reads translator metadata and loads translators from a local directory
+ * where translator NPM packages have been installed.
+ */
+export class LocalPackageSource extends PackageSource {
+    /**
+     * Gets all subdirectories (not files) under a directory.
+     */
+    private static async getSubdirectoriesAsync(directoryPath: string): Promise<string[]> {
+        let directoryNames: string[] = [];
+        let childNames: string[] = await fs.readdir(directoryPath);
+
+        for (let i = 0; i < childNames.length; i++) {
+            if ((await fs.stat(path.join(directoryPath, childNames[i]))).isDirectory()) {
+                directoryNames.push(childNames[i]);
+            }
+        }
+
+        return directoryNames;
+    }
+
+    /**
+     * Path to the directory where packages are cached. In order to enable requiring
+     * modules in these packages without using relative paths, the path should be a
+     * node_modules directory that is included in the search path.
+     */
+    public readonly cacheDirectory: string;
```

Comment:
Seems reasonable to keep this readonly property public.

But note all the `PackageSource` subclasses including `LocalPackageSource` class are intended to be sort of private, in the sense that they aren't exported from index.ts and won't be projected into the client APIs. Callers are expected to use the factory methods on `PackageSource` to create them, then invoke methods via the base class. The primary reason for this is to reduce the surface area of projected APIs.

---
In reply to: [73274368](https://github.com/openT2T/opent2t/pull/2#discussion_r73274368) [](ancestors = 73274368)